### PR TITLE
build: pack charm in provided environment when using bases (CRAFT-313)

### PR DESCRIPTION
**Pull request by cjp256**
For each bases configuration, check if a build-on base can be
provided.  If so, launch environment and execute pack command
inside each to build charm for each viable bases config index.

Future work should extend the pack command to include the parent's
verbose or quiet mode flags for consistency.

Signed-off-by: Chris Patterson <chris.patterson@canonical.com>

### CODE REVIEW

File: charmcraft/commands/build.py [hunk 1]
```diff
@@ -191,14 +193,29 @@ def run(self, bases_indices: Optional[List[int]] = None) -> List[str]:
...
                         )
-                        charm_name = self.build_charm(bases_config)
+                        if is_charmcraft_running_in_managed_mode():
+                            charm_name = self.build_charm(bases_config)
+                        else:
+                            with launched_environment(
+                                charm_name=self.metadata.name,
+                                project_path=self.charmdir,
+                                base=build_on,
+                                bases_index=i,
```
Comment (sergiusens) [inline charmcraft/commands/build.py:214]: personal preference comment, and I think I made it in the past in a different PR, but i and j should only be used for simple iterations, it's still used 10 lines below and hard to keep track of

File: charmcraft/commands/build.py [hunk 2]
```diff
@@ -225,6 +242,22 @@ def run(self, bases_indices: Optional[List[int]] = None) -> List[str]:
 
         return charms
 
+    def build_charm_in_instance(self, instance: Executor, bases_index: int) -> str:
```
Comment (sergiusens) [inline charmcraft/commands/build.py:245]: I would use a name different than build to not confuse in the future when parts lifecycle is introduced

File: charmcraft/commands/build.py [hunk 3]
```diff
@@ -177,11 +183,7 @@ def run(self, bases_indices: Optional[List[int]] = None) -> List[str]:
         """
         charms: List[str] = []
 
-        if is_charmcraft_running_in_managed_mode():
-            if self.config.bases is None:
-                # XXX: 2021-06-16 Patterson temporary until we have the default base
-                raise CommandError("Bases are currently required in managed-mode.")
-
+        if self.config.bases:
             for i, bases_config in enumerate(self.config.bases):
```
Comment (sergiusens) [inline charmcraft/commands/build.py:187]: given the size of this, can we separate selection/planning from execution?

File: charmcraft/commands/build.py (hunk 1 above)
Comment (cjp256) [reply charmcraft/commands/build.py:214]: Good point, will fix!

File: charmcraft/commands/build.py (hunk 2 above)
Comment (cjp256) [reply charmcraft/commands/build.py:245]: How about s/build/pack?

File: charmcraft/commands/build.py (hunk 3 above)
Comment (cjp256) [reply charmcraft/commands/build.py:187]: Sounds good, will do!

File: charmcraft/commands/build.py (hunk 1 above)
Comment (cjp256) [reply charmcraft/commands/build.py]: Done

File: charmcraft/commands/build.py (hunk 2 above)
Comment (cjp256) [reply charmcraft/commands/build.py]: Done, renamed to pack_charm_in_instance()

File: charmcraft/commands/build.py [hunk 4]
```diff
@@ -177,41 +183,54 @@ def run(self, bases_indices: Optional[List[int]] = None) -> List[str]:
...
         charms: List[str] = []
 
-        if is_charmcraft_running_in_managed_mode():
-            if self.config.bases is None:
-                # XXX: 2021-06-16 Patterson temporary until we have the default base
-                raise CommandError("Bases are currently required in managed-mode.")
-
-            for i, bases_config in enumerate(self.config.bases):
-                if bases_indices and i not in bases_indices:
+        if self.config.bases:
```
Comment (facundobatista) [inline charmcraft/commands/build.py:186]: Shouldn't we always have "bases"? Even if not mandatory yet, some default?

Comment (cjp256) [reply charmcraft/commands/build.py:186]: Yes.  But that PR is upcoming.  At the moment this keeps behavior consistent.

**Review by facundobatista — APPROVED** [review]
Great, thanks!

**Review by sergiusens — APPROVED** [review]
Looks better, but I don't see my comment https://github.com/canonical/charmcraft/pull/411#discussion_r659027840 addressed. It's fine if it is a different PR.

**Pull request merged**